test-network-k8s/scripts/chaincode.sh

```diff
@@ -69,30 +69,66 @@ function chaincode_command_group() {
 # Convenience routine to "do everything" required to bring up a sample CC.
 function deploy_chaincode() {
   local cc_name=$1
-  local cc_label=$2
-  local cc_folder=$(absolute_path $3)
-
+  local cc_label=$1
+  local cc_folder=$(absolute_path $2)
   local temp_folder=$(mktemp -d)
   local cc_package=${temp_folder}/${cc_name}.tgz
 
-  package_chaincode       ${cc_label} ${cc_name} ${cc_package}
+  prepare_chaincode_image ${cc_folder} ${cc_name}
+  package_chaincode       ${cc_name} ${cc_label} ${cc_package}
+
+  if [ "${CHAINCODE_BUILDER}" == "ccaas" ]; then
+    set_chaincode_id      ${cc_package}
+    launch_chaincode      ${cc_name} ${CHAINCODE_ID} ${CHAINCODE_IMAGE}
+  fi
+
+  activate_chaincode      ${cc_name} ${cc_package}
+}
 
-  set_chaincode_id        ${cc_package}
-  set_chaincode_image     ${cc_folder}
+# Prepare a chaincode image for use in a builder package.
+# Sets the CHAINCODE_IMAGE environment variable
+function prepare_chaincode_image() {
+  local cc_folder=$1
+  local cc_name=$2
 
-  build_chaincode_image   ${cc_folder} ${CHAINCODE_IMAGE}
+  build_chaincode_image ${cc_folder} ${cc_name}
 
-  if [ "${CLUSTER_RUNTIME}" == "kind" ]; then
-    kind_load_image       ${CHAINCODE_IMAGE}
+  if [ "${CLUSTER_RUNTIME}" == "k3s" ]; then
+    # For rancher / k3s runtimes, bypass the local container registry and load images directly from the image cache.
+    export CHAINCODE_IMAGE=${cc_name}
+  else
+    # For KIND and k8s-builder environments, publish the image to a local docker registry
+    export CHAINCODE_IMAGE=localhost:${LOCAL_REGISTRY_PORT}/${cc_name}
+    publish_chaincode_image ${cc_name} ${CHAINCODE_IMAGE}
   fi
+}
 
-  launch_chaincode        ${cc_name} ${CHAINCODE_ID} ${CHAINCODE_IMAGE}
-  activate_chaincode      ${cc_name} ${cc_package}
+function build_chaincode_image() {
+  local cc_folder=$1
+  local cc_name=$2
+
+  push_fn "Building chaincode image ${cc_name}"
+
+  $CONTAINER_CLI build ${CONTAINER_NAMESPACE} -t ${cc_name} ${cc_folder}
+
+  pop_fn
+}
+
+# tag a docker image with a new name and publish to a remote container registry
+function publish_chaincode_image() {
+  local cc_name=$1
+  local cc_url=$2
+  push_fn "Publishing chaincode image ${cc_url}"
+
+  ${CONTAINER_CLI} tag  ${cc_name} ${cc_url}
+  ${CONTAINER_CLI} push ${cc_url}
+
+  pop_fn
 }
 
 # Infer a reasonable name for the chaincode image based on the folder path conventions, or
 # allow the user to override with TEST_NETWORK_CHAINCODE_IMAGE.
-function set_chaincode_image() {
+function zz_unused_set_chaincode_image() {
```

Comment:
Did you mean to leave this in?

Reply:
Whoops !   That one almost survived the chopping block.   At least the function name was clear ....